### PR TITLE
Check flags that are not supposed to be mutated by completion

### PR DESCRIPTION
**Pull request by anatoliykmetyuk**

### CODE REVIEW

**Comment by smarter** [comment]
There's some flags for which this isn't true currently: https://github.com/lampepfl/dotty/issues/4221
I think the best place to have these checks is in `setFlag` and `resetFlag`, these methods should have assertions somewhat similar to https://github.com/lampepfl/dotty/blob/2e2a2df963dbb4c0fce9c162f0f579834a173ed2/compiler/src/dotty/tools/dotc/core/SymDenotations.scala#L552-L554 and https://github.com/lampepfl/dotty/blob/2e2a2df963dbb4c0fce9c162f0f579834a173ed2/compiler/src/dotty/tools/dotc/core/SymDenotations.scala#L306-L308

**Comment by anatoliykmetyuk** [comment]
Test performance please

**Comment by anatoliykmetyuk** [comment]
On top #7449.

**Comment by anatoliykmetyuk** [comment]
Test performance please

**Comment by anatoliykmetyuk** [comment]
Was that the title case letter or am I not in the whitelist? AFAIK I was added to the bot's whitelist at some point...

**Comment by liufengyun** [comment]
Yes, it seems to be case-sensitive, we should improve:

https://github.com/lampepfl/bench/blob/master/bin/process#L45

**Comment by anatoliykmetyuk** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/core/SymDenotations.scala [hunk 1]
```diff
@@ -160,20 +160,34 @@ object SymDenotations {
      */
     private[dotc] final def flagsUNSAFE: FlagSet = myFlags
 
+    private def setMyFlags(flags: FlagSet) =
+      lazy val immutableFlags = if myInfo.isInstanceOf[SymbolLoader] then AfterLoadFlags else FromStartFlags
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/SymDenotations.scala]: Every local lazy val is a boxed object, should be avoided in performance sensitive code like this.

**Comment by anatoliykmetyuk** [comment]
> Spending a flag on an internal check is too wasteful, I think.

True, only 2 flags left. How bad will the performance impact be if we add another `Long` `var` giving us 64 more flags? If you look at the performance test charts, the three latest data points are (in chronological order):

- This PR where I used a `Boolean` `var` instead of the `Loading` flag.
- Unrelated PR
- This PR where I replaced the dedicated var with a flag.

I assumed this would help performance – however, it seems it only positively impacted the performance of `dotty (source changes over time)` test. Other tests are roughly at the same level for the var and the flag versions. Which may mean adding another `var` to the `SymDenotation` may not cause as much of a performance impact? Or can we just fall back to using a boolean var instead of a flag?

> What is the reason for introducing a new flag

This is because we have the `AfterLoadFlags` which are immutable after a symbol is loaded (on load time, we can modify whatever flags we want). However, loading happens on completion of symbols, and on completion we are not allowed to modify certain flags. So the rules of flag mutability are as follows:

- If we are on completion, do not modify `FromStartFlags`
- If we are on completion but loading the symbol, allow modification of all the flags
- If we are on completion after loading, do not modify `AfterLoadFlags`

So there are three different states to track here which requires two bits. The following chunks of code are considered as symbol loading (between `startedLoading()` and `finishedLoading()` method calls).

- Loading of symbols from classfiles: https://github.com/lampepfl/dotty/pull/7427/files#diff-ef84bd1ce9ffda978ae9ae3426c1cf1dR396-R409
- The late compilation, as far as I understand this is for loading symbols from the Java sources: https://github.com/lampepfl/dotty/pull/7427/files#diff-ef84bd1ce9ffda978ae9ae3426c1cf1dR419-R421
- When we replace an existing symbol from the tree unpickler: https://github.com/lampepfl/dotty/pull/7427/files#diff-8865435e54a057041718570580f5d24fR544-R552
- When unpickling Scala 2 trees: https://github.com/lampepfl/dotty/pull/7427/files#diff-59330d0079b0029a4e62c5d5d0385db2R194-R231

> Also, the code used in hotspots is too expensive.

Is there anything else apart from the lazy vals that box and the `ConditionallyImmutableFlags` scaffolding?

> I believe this suffers from too early generalization. It seems to the problem is only with Deferred, right? What exactly is the problem? Is there a cheaper way to solve it, without creating an expensive general scaffolding?

The problem with the `Deferred` is that the mechanism of the opaque types sets this flag for opaque types on their completion. Hence we cannot assume that `Deferred` is a `FromStartFlag`. However, if we just drop this assumption, things go wrong on unpickling at this line:

https://github.com/lampepfl/dotty/blob/d2db509f7b47c16edcad7d0772c58826c6cde897/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala#L675

So this line queries the flags of some class member `sym` while unpickling that class. `!sym.isOneOf(DeferredOrLazyOrMethod))`. If we try to complete the `sym` at that place, we'll get an exception like `no symbol at Addr(23)` – because the parent class was not fully unpickled when we tried unpickling its member symbol. However, if we cannot assume `Deferred` is immutable during completion, we have no other choice but to complete `sym` on that call – otherwise we cannot guarantee the symbol's `Deferred` status won't change.

So `Deferred` cannot be assumed immutable on completion but this assumption is critical for the correct unpickling order.

At the same time, at that particular place we can be absolutely sure `Deferred` will not change after completion. This is because `isOneOf` is evaluated only if `sym.isTerm` is true (the first operand to the `&&`). And we know that `Deferred` is mutated only for opaque types, which are types and not terms.

Hence the solution of the conditionally immutable flags.

> Is there a cheaper way to solve it, without creating an expensive general scaffolding?

We can just hard-code the `Deferred` flag in the `isCurrent` method.

**Comment by odersky** [comment]
Can we set Deferred for opaque types from the start? If not, then the next best choice is to hardcode Deferred in isCurrent.

> Is there anything else apart from the lazy vals that box and the ConditionallyImmutableFlags scaffolding?

I believe so. A reasonable way to track this is to count how many object allocations are executed by the code (lambdas count). For hot code the goal is 0.

> If we are on completion but loading the symbol, allow modification of all the flags

Why do we need this?

**Comment by anatoliykmetyuk** [comment]
> Why do we need this?

I believe because we are creating a symbol from scratch, so we should be able to modify anything at all about it.

**Comment by anatoliykmetyuk** [comment]
test performance please

**Pull request closed without merging**